akka-stream/src/main/scala/akka/stream/impl/fusing/StreamOfStreams.scala

```diff
@@ -212,6 +212,11 @@ import scala.collection.JavaConverters._
   override def toString: String = s"PrefixAndTail($n)"
 }
 
+private[akka] object GroupBy {
+  val TooManySubstreamsOpen =
```

Comment:
Since this is exposed to end users it shouldn't be internal. Define the exception in same way as for example: https://github.com/akka/akka/blob/master/akka-stream/src/main/scala/akka/stream/StreamDetachedException.scala

Since it's exposed in both scaladsl and javadsl and we want the only one exception type it must be in `akka.stream`.

You can still keep this `val` as the single instance that is thrown

Reply:
Since it will anyway be a different class I think that a real class is better:
```
TooManySubstreamsOpenException extends IllegalStateException(...) with NoStackTrace
```

Reply:
Wilco. I’ll do it first thing my time.

Reply:
@patriknw I've updated my PR as per your requests. It should remain backwardly compatible while also having introduced a more formal exception.